packages/ui-extensions-react/src/surfaces/admin/render.tsx

```diff
@@ -36,17 +36,15 @@ export function reactExtension<ExtensionTarget extends RenderExtensionTarget>(
   return extension<'Playground'>(target as any, async (root, api) => {
     const element = await render(api as ApiForRenderExtension<ExtensionTarget>);
 
-    await new Promise<void>((resolve, reject) => {
+    return new Promise<() => void>((resolve, reject) => {
       try {
-        remoteRender(
+        const reactRoot = createRoot(root);
+        reactRoot.render(
           <ExtensionApiContext.Provider value={api}>
             {element}
           </ExtensionApiContext.Provider>,
-          root,
-          () => {
-            resolve();
-          },
         );
+        resolve(() => reactRoot.unmount());
```

Comment:
So, we decide to return a clean up function for UI extension and just data object for Data extension then.

Reply:
I'm leaning toward this but we haven't decided. This is being discussed in this issue: https://github.com/Shopify/ui-extensions/pull/2353#issuecomment-2426849901